File: pylimerpredictor/predictors/nn_predictor.py
#!/usr/bin/env python
import io
import os
import pickle
from typing import List

from termcolor import colored
import torch

from pylimerpredictor.models.prediction_input import PredictionInput

from .neural_network.config import fit_key, input_axes, output_axes
from .neural_network.custom_scaler import CustomScaler, TargetScaler
from .neural_network.my_nn import NeuralNetwork


class RenameUnpickler(pickle.Unpickler):
    def find_class(self, module, name):
        renamed_module = module
        if module == "custom_scaler":
            renamed_module = "pylimerpredictor.predictors.neural_network.custom_scaler"

        return super(RenameUnpickler, self).find_class(renamed_module, name)


def renamed_load(file_obj):
    return RenameUnpickler(file_obj).load()


def renamed_loads(pickled_bytes):
    file_obj = io.BytesIO(pickled_bytes)
    return renamed_load(file_obj)


def predict_nn_results(prediction_input: PredictionInput) -> dict:
    base_path = os.path.dirname(__file__)
    model_path = os.path.join(
        base_path, "neural_network", "final_model_{}.pth".format(fit_key)
    )
    scaler_path = os.path.join(
        base_path, "neural_network", "scaler_{}.pkl".format(fit_key)
    )

    model = NeuralNetwork()
    # load the trained model
    model.load_state_dict(
        torch.load(
            model_path,
            weights_only=True,
        )
    )
    model.eval()

    # Load the saved scaler
    with open(scaler_path, "rb") as f:
        scalers = renamed_load(f)  # pickle.load(f)
        scaler = scalers["standard"]
        assert isinstance(scaler, CustomScaler), "Expected CustomScaler instance"
        target_scaler = scalers["target"]
        assert isinstance(target_scaler, TargetScaler), "Expected TargetScaler instance"

    # Function to make a prediction
    def predict(input_values: List):
        input_tensor = scaler.transform([input_values])
        input_tensor = torch.tensor(input_tensor, dtype=torch.float32)
        with torch.no_grad():
            output = model(input_tensor)
        return target_scaler.inverse_transform(output.numpy())[0]

    input_value_translator = {
        "r": prediction_input.stoichiometric_imbalance,
        "p": prediction_input.crosslink_conversion,
        "b2": prediction_input.get_b2(),
        "ge_1 [MPa]": prediction_input.plateau_modulus.to("MPa").magnitude,
        "temperature [K]": prediction_input.temperature.to("K").magnitude,
        "density [g/cm^3]": prediction_input.density.to("g/cm^3").magnitude,
        "param's <b> [nm]": prediction_input.get_mean_bead_distance()
        .to("nm")
        .magnitude,
        "param's Mw [kg/mol]": prediction_input.bead_mass.to("kg/mol").magnitude,
        "Mw [kg/mol]": prediction_input.n_beads_bifunctional
        * prediction_input.bead_mass.to("kg/mol").magnitude,
        "Mw [kg/mol] monofunctional chains": prediction_input.n_beads_monofunctional
        * prediction_input.bead_mass.to("kg/mol").magnitude,
        "Mw [kg/mol] xlink chains": prediction_input.n_beads_xlinks
        * prediction_input.bead_mass.to("kg/mol").magnitude,
        "Mw [kg/mol] solvent chains": prediction_input.n_beads_zerofunctional
        * prediction_input.bead_mass.to("kg/mol").magnitude,
        "functionality_per_xlink_chain": int(prediction_input.crosslink_functionality),
        "functionalize_discrete": bool(prediction_input.functionalize_discrete),
        "remove_wsol": bool(prediction_input.extract_solvent_before_measurement),
        "solvent_fraction_of_beads": prediction_input.get_total_n_beads_solvent()
        / prediction_input.get_n_total_beads(),
        "monofunctional_fraction_of_beads": prediction_input.get_total_n_beads_monofunctional()
        / prediction_input.get_n_total_beads(),
        "bifunctional_fraction_of_beads": prediction_input.get_total_n_beads_bifunctional()
        / prediction_input.get_n_total_beads(),
        "crosslink_fraction_of_beads": prediction_input.get_total_n_beads_xlinks()
        / prediction_input.get_n_total_beads(),
        "entanglement_sampling_cutoff [nm]": prediction_input.entanglement_sampling_cutoff.to(
            "nm"
        ).magnitude,
    }

    assert all(label in input_value_translator for label in input_axes)
    predictions = predict([input_value_translator[k] for k in input_axes])

    print(colored("NN Predictions: {}".format(predictions), "green"))

    result_translator = {
        "g_eq": "G [MPa] from gamma Mean, Entangled FB No Slipping",
        "g_phantom": "Phantom, FB/FR, G_ANT [MPa]",
        "w_dangling": "dangling_fraction Mean, Entangled FB No Slipping",
        "w_soluble": "soluble_fraction Mean, Entangled FB No Slipping",
    }

    results = {}

    assert any(
        label in output_axes for label in result_translator.values()
    ), "No results to return, check the output axes and result translator."

    for key, label in result_translator.items():
        if label in output_axes:
            index = output_axes.index(label)
            results[key] = predictions[index]
        else:
            results[key] = None

    if "g_eq" in results and "g_phantom" in results:
        results["g_entangled"] = results["g_eq"] - results["g_phantom"]

    if "w_dangling" in results and "w_soluble" in results:
        results["w_backbone"] = 1 - results["w_dangling"] - results["w_soluble"]

    return results

File: src/worker_predictor/nn/nn_predictor.ts
import type { PredictionInput } from "../../entity/PredictionInput";
import { ModulusPredictionOutput } from "../../entity/PredictionOutput";
import type { Predictor } from "../predictor";
import Qty from "js-quantities";
import * as ort from "onnxruntime-web";
import { CustomScaler, TargetScaler, type ModelMetadata } from "./scaler";

interface ModelConfig {
  modelPath: string;
  metadataPath: string;
}

/**
 * Neural Network Predictor that automatically selects the optimal model based on input parameters.
 *
 * This predictor manages three different ONNX models:
 * - General fallback model (m-): Handles all polymer types and configurations (20 features)
 * - PDMS long-xlinks model: Optimized for PDMS with multiple crosslinks (9 features)
 * - PDMS 1xlink-without-mono model: Optimized for PDMS with single crosslinks and no monofunctional chains (6 features)
 *
 * The predictor uses caching to avoid reloading models and maintains separate scalers for each model.
 * Models are selected automatically based on polymer type and structural parameters for optimal accuracy.
 */
export class NNPredictor implements Predictor {
  // Cache for loaded models and metadata
  private modelCache = new Map<string, ort.InferenceSession>();
  private metadataCache = new Map<string, ModelMetadata>();
  private scalerCache = new Map<
    string,
    { input: CustomScaler; target: TargetScaler }
  >();

  // Current model configuration
  private currentConfig: ModelConfig | null = null;

  constructor() {
    // No longer hardcode model paths - they will be selected dynamically
  }

  /**
   * Selects the appropriate neural network model based on input parameters.
   *
   * The model selection logic prioritizes specialized models for PDMS polymers:
   * 1. Use 1xlink-without-mono model for PDMS with single crosslinks and no monofunctional chains
   * 2. Use long-xlinks model for PDMS with multiple crosslinks
   * 3. Use general fallback model for all other cases (non-PDMS or other configurations)
   *
   * @param input - The prediction input containing polymer parameters
   * @returns ModelConfig object with paths to the selected model and metadata files
   */
  private selectModel(input: PredictionInput): ModelConfig {
    // Check if it's PDMS polymer type
    const isPDMS = input.polymer_name.toLowerCase().includes("pdms");

    if (
      isPDMS &&
      !input.extract_solvent_before_measurement &&
      !input.functionalize_discrete &&
      0 == input.n_zerofunctional_chains
    ) {
      // For PDMS: n_beads_xlinks = 1 and no monofunctional chains
      // Uses specialized model trained on subset with fewer parameters for better performance
      if (input.n_beads_xlinks === 1 && input.n_monofunctional_chains === 0) {
        return {
          modelPath:
            "/models/m-remove_wsol_False_entanglements_as_springs_False_has_solvent_False_polymer_type_pdms_architecture_1xlink-without-mono.onnx",
          metadataPath:
            "/models/m-remove_wsol_False_entanglements_as_springs_False_has_solvent_False_polymer_type_pdms_architecture_1xlink-without-mono_metadata.json",
        };
      }
      // For PDMS: n_beads_xlinks > 1
      // Uses specialized model for long crosslink configurations
      else if (input.n_beads_xlinks > 1) {
        return {
          modelPath:
            "/models/m-remove_wsol_False_functionalize_discrete_False_entanglements_as_springs_False_has_solvent_False_polymer_type_pdms_architecture_long-xlinks.onnx",
          metadataPath:
            "/models/m-remove_wsol_False_functionalize_discrete_False_entanglements_as_springs_False_has_solvent_False_polymer_type_pdms_architecture_long-xlinks_metadata.json",
        };
      }
    }

    // Fallback to general model for non-PDMS polymers or other PDMS configurations
    return {
      modelPath: "/models/m-.onnx",
      metadataPath: "/models/m-_metadata.json",
    };
  }

  private async loadModel(config: ModelConfig): Promise<ort.InferenceSession> {
    // Check if model is already cached
    if (!this.modelCache.has(config.modelPath)) {
      try {
        const session = await ort.InferenceSession.create(config.modelPath);
        this.modelCache.set(config.modelPath, session);
      } catch (error) {
        throw new Error(
          `Failed to load ONNX model from ${config.modelPath}: ${error}`
        );
      }
    }
    return this.modelCache.get(config.modelPath)!;
  }

  private async loadMetadata(config: ModelConfig): Promise<ModelMetadata> {
    // Check if metadata is already cached
    if (!this.metadataCache.has(config.metadataPath)) {
      try {
        const response = await fetch(config.metadataPath);
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`);
        }
        const metadata = (await response.json()) as ModelMetadata;
        this.metadataCache.set(config.metadataPath, metadata);

        // Initialize and cache scalers
        const inputScaler = new CustomScaler(metadata.input_scaler);
        const targetScaler = new TargetScaler(metadata.target_scaler);
        this.scalerCache.set(config.metadataPath, {
          input: inputScaler,
          target: targetScaler,
        });
      } catch (error) {
        throw new Error(
          `Failed to load model metadata from ${config.metadataPath}: ${error}`
        );
      }
    }
    return this.metadataCache.get(config.metadataPath)!;
  }

  private createFeatureMapping(): Map<
    string,
    (input: PredictionInput) => number
  > {
    // Map from NN input names to functions that extract the corresponding value from PredictionInput
    return new Map([
      ["r", (input) => input.stoichiometric_imbalance],
      ["p", (input) => input.crosslink_conversion],
      ["b2", (input) => input.b2],
      ["ge_1 [MPa]", (input) => input.plateau_modulus.to("MPa").scalar],
      ["temperature [K]", (input) => input.temperature.to("kelvin").scalar],
      ["density [g/cm^3]", (input) => input.density.to("g/cm^3").scalar],
      ["param's <b> [nm]", (input) => input.mean_bead_distance.to("nm").scalar],
      [
        "param's <b^2> [nm^2]",
        (input) => input.mean_squared_bead_distance.to("nm^2").scalar,
      ],
      [
        "remove_wsol",
        (input) => (input.extract_solvent_before_measurement ? 1 : 0),
      ],
      ["param's Mw [kg/mol]", (input) => input.bead_mass.to("kg/mol").scalar],
      [
        "Mw [kg/mol]",
        (input) =>
          input.bead_mass.to("kg/mol").scalar * input.n_beads_bifunctional,
      ],
      [
        "Mw [kg/mol] monofunctional chains",
        (input) =>
          input.bead_mass.to("kg/mol").scalar * input.n_beads_monofunctional,
      ],
      [
        "Mw [kg/mol] solvent chains",
        (input) =>
          input.bead_mass.to("kg/mol").scalar * input.n_beads_zerofunctional,
      ],
      [
        "Mw [kg/mol] xlink chains",
        (input) => input.bead_mass.to("kg/mol").scalar * input.n_beads_xlinks,
      ],
      [
        "functionality_per_xlink_chain",
        (input) => input.crosslink_functionality,
      ],
      [
        "functionalize_discrete",
        (input) => (input.functionalize_discrete ? 1 : 0),
      ],
      [
        "solvent_fraction_of_beads",
        (input) => input.total_n_beads_solvent / input.n_beads_total,
      ],
      [
        "monofunctional_fraction_of_beads",
        (input) => input.total_n_beads_monofunctional / input.n_beads_total,
      ],
      [
        "bifunctional_fraction_of_beads",
        (input) => input.total_n_beads_bifunctional / input.n_beads_total,
      ],
      [
        "entanglement_sampling_cutoff [nm]",
        (input) => input.entanglement_sampling_cutoff.to("nm").scalar,
      ],
      [
        "n_chains_total",
        (input) =>
          input.n_bifunctional_chains +
          input.n_monofunctional_chains +
          input.n_zerofunctional_chains +
          input.n_chains_crosslinks,
      ],
      ["entanglements_as_springs", (input) => 0],
    ]);
  }

  private async prepareInput(
    input: PredictionInput,
    config: ModelConfig
  ): Promise<Record<string, ort.Tensor>> {
    // Load metadata to get feature order
    const metadata = await this.loadMetadata(config);

    // Get cached scalers
    const scalers = this.scalerCache.get(config.metadataPath);
    if (!scalers) {
      throw new Error("Input scaler not initialized");
    }

    // Get the feature mapping
    const featureMapping = this.createFeatureMapping();

    // Build features array based on the metadata's input_features order
    const features: number[] = [];

    for (const featureName of metadata.input_features) {
      const extractFunction = featureMapping.get(featureName);
      if (extractFunction) {
        features.push(extractFunction(input));
      } else {
        throw new Error(
          `Unknown input feature: ${featureName}. Available mappings: ${Array.from(
            featureMapping.keys()
          ).join(", ")}`
        );
      }
    }

    // Apply input scaling
    const scaledFeatures = scalers.input.transform(features);

    // Create input tensor with the correct feature order
    const inputTensor = new ort.Tensor(
      "float32",
      new Float32Array(scaledFeatures),
      [1, scaledFeatures.length]
    );

    // Load the model to get the correct input name
    const session = await this.loadModel(config);
    const inputName = session.inputNames[0]; // Use first (and likely only) input name

    return { [inputName]: inputTensor };
  }

  private processOutput(
    outputs: ort.InferenceSession.OnnxValueMapType,
    config: ModelConfig
  ): ModulusPredictionOutput {
    // Get cached scalers
    const scalers = this.scalerCache.get(config.metadataPath);
    if (!scalers) {
      throw new Error("Target scaler not initialized");
    }

    // Extract the output tensor
    const outputNames = Object.keys(outputs);
    if (outputNames.length === 0) {
      throw new Error("No outputs received from ONNX model");
    }

    const outputTensor = outputs[outputNames[0]] as ort.Tensor;
    const outputData = outputTensor.data as Float32Array;

    // Convert to regular array and apply inverse target scaling
    const scaledOutput = Array.from(outputData);
    const finalOutput = scalers.target.inverseTransform(scaledOutput);

    // Assuming the model outputs 4 values based on metadata output_features:
    // ["G [MPa] from gamma Mean, Entangled FB No Slipping", "Phantom, FB/FR, G_ANT [MPa]", "dangling_fraction Mean, Entangled FB No Slipping", "soluble_fraction Mean, Entangled FB No Slipping"]
    if (finalOutput.length < 4) {
      throw new Error(
        `Expected at least 4 output values, got ${finalOutput.length}`
      );
    }

    const result = new ModulusPredictionOutput();
    result.entanglement_modulus = Qty(finalOutput[0], "MPa"); // G [MPa] from gamma Mean, Entangled FB No Slipping
    result.phantom_modulus = Qty(finalOutput[1], "MPa"); // Phantom, FB/FR, G_ANT [MPa]
    result.w_dangling = Qty(finalOutput[2], ""); // dangling_fraction Mean, Entangled FB No Slipping
    result.w_soluble = Qty(finalOutput[3], ""); // soluble_fraction Mean, Entangled FB No Slipping

    return result;
  }

  async predict(input: PredictionInput): Promise<ModulusPredictionOutput> {
    try {
      // Select the appropriate model based on input parameters
      const config = this.selectModel(input);
      this.currentConfig = config;

      // Load the selected model and metadata
      const session = await this.loadModel(config);
      await this.loadMetadata(config); // Ensure metadata and scalers are loaded

      // Prepare input data (includes scaling)
      const feeds = await this.prepareInput(input, config);

      // Run inference
      const results = await session.run(feeds);

      // Process and return output (includes inverse scaling)
      return this.processOutput(results, config);
    } catch (error) {
      console.error("NN prediction error:", error);
      throw new Error(`Neural network prediction failed: ${error}`);
    }
  }
}
